server/entity/area_effect_cloud.go

```diff
@@ -0,0 +1,240 @@
+package entity
+
+import (
+	"github.com/df-mc/dragonfly/server/block/cube"
+	"github.com/df-mc/dragonfly/server/entity/effect"
+	"github.com/df-mc/dragonfly/server/internal/nbtconv"
+	"github.com/df-mc/dragonfly/server/item/potion"
+	"github.com/df-mc/dragonfly/server/world"
+	"github.com/go-gl/mathgl/mgl64"
+	"time"
+)
+
+// AreaEffectCloud is the cloud that is created when: lingering potions are thrown; creepers with potion effects explode;
+// dragon fireballs hit the ground.
+type AreaEffectCloud struct {
+	duration           int64
+	reapplicationDelay int64
+	durationOnUse      int64
+
+	radius       float64
+	radiusOnUse  float64
+	radiusGrowth float64
+
+	close bool
+
+	targets map[world.Entity]int64
+
+	t   potion.Potion
+	age int64
+
+	transform
+}
+
+// NewDefaultAreaEffectCloud ...
+func NewDefaultAreaEffectCloud(pos mgl64.Vec3, t potion.Potion) *AreaEffectCloud {
+	return NewAreaEffectCloud(pos, t, time.Second*30, time.Second*2, 0, 3.0, -0.5, -0.005)
+}
+
+// NewAreaEffectCloud ...
+func NewAreaEffectCloud(pos mgl64.Vec3, t potion.Potion, duration, reapplicationDelay, durationOnUse time.Duration, radius, radiusOnUse, radiusGrowth float64) *AreaEffectCloud {
+	a := &AreaEffectCloud{
+		duration:           duration.Milliseconds() / 50,
+		reapplicationDelay: reapplicationDelay.Milliseconds() / 50,
+		durationOnUse:      durationOnUse.Milliseconds() / 50,
+
+		radius:       radius,
+		radiusOnUse:  radiusOnUse,
+		radiusGrowth: radiusGrowth,
+
+		targets: make(map[world.Entity]int64),
+		t:       t,
+	}
+	a.transform = newTransform(a, pos)
+	return a
+}
+
+// Name ...
+func (a *AreaEffectCloud) Name() string {
+	return "Area Effect Cloud"
+}
+
+// EncodeEntity ...
+func (a *AreaEffectCloud) EncodeEntity() string {
+	return "minecraft:area_effect_cloud"
+}
+
+// Duration returns the duration of the cloud.
+func (a *AreaEffectCloud) Duration() time.Duration {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	return time.Duration(a.duration) * time.Millisecond * 50
+}
+
+// Radius returns information about the cloud's, radius, change rate, and growth rate.
+func (a *AreaEffectCloud) Radius() (radius, radiusOnUse, radiusGrowth float64) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	return a.radius, a.radiusOnUse, a.radiusGrowth
+}
+
+// Effects returns the effects the area effect cloud provides.
+func (a *AreaEffectCloud) Effects() []effect.Effect {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	return a.t.Effects()
+}
+
+// BBox ...
+func (a *AreaEffectCloud) BBox() cube.BBox {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	return cube.Box(-a.radius, 0, -a.radius, a.radius, 0.5, a.radius)
+}
+
+// Tick ...
+func (a *AreaEffectCloud) Tick(w *world.World, _ int64) {
```

Comment:
Really don't like how this method can lock/unlock at least 3 times, plus 2*the amount of entities in radius

Reply:
I think it's better now? I also removed the locks/unlocks for each entity.